_posts/2021-04-05-fmf-unified-testing.md

```diff
@@ -0,0 +1,156 @@
+
+```yaml
+- hosts: localhost
+  roles:
+  - role: standard-test-source
+    tags:
+    - always
+
+  - role: standard-test-basic
+    tags:
+    - classic
+    required_packages:
+    - cockpit
+    [...]
+    - npm
+    - python3
+    tests:
+    - verify:
+        dir: .
+        run: ./verify.sh
+        save-files: ["logs/*"]
+```
+
+Aside from that you just need some boilerplate `.fmf/version` (just "1") and at least one top-level [plans/*.fmf](https://github.com/cockpit-project/starter-kit/blob/master/plans/all.fmf) (which can really be the same for every project, and hopefully it will be the implied default some day).
+
+This test metadata format provides the underpinnings for the following new tools.
+
+## Test Management Tool
+
+The relatively new [Test Management Tool](https://docs.fedoraproject.org/en-US/ci/tmt/) (`tmt`) addresses the first two points above: Provided that you project has FMF metadata for its tests, a simple
+
+    tmt run
+
+downloads a standard Fedora development series VM (34 at the moment), starts it in libvirt/QEMU, runs your tests inside it, gets you a live log while the test is running, copies out all the test logs and artifacts, and cleans up everything again. You can customize this in various dimensions, such as running on a different Fedora release:
+
+    tmt run --all provision --how virtual --image fedora-33
+
+or run the steps only until the `report` stage (thus skipping `finish`), so that you can ssh into the test VM and investigate failures:
+
+    tmt run --until report
+
+See `--help` and the [documentation](https://docs.fedoraproject.org/en-US/ci/tmt/#_execute_tests) for details.
+
+Until recently this only worked with `qemu:///system` libvirt, i.e. not in containers or [toolbox](https://docs.fedoraproject.org/en-US/fedora-silverblue/toolbox/). Fortunately, the latest [testcloud](https://pagure.io/testcloud/pull-request/83) and [tmt](https://github.com/psss/tmt/issues/541) versions now switched to `qemu:///session` by default (thanks to Petr Šplíchal for responding to my nagging so quickly!), so that `tmt` now makes *zero* root-only configuration assumptions about the hosts and does not need any root privileges, bridges, or services.
+
+## Packit
+
+[Packit](https://packit.dev/) started out as a tool and service to automatically package your project's new upstream releases into Fedora or COPR. It recently learned a really cool new trick, though: The [Packit-as-a-Service GitHub app](https://github.com/marketplace/packit-as-a-service) runs your project's FMF test plans in upstream pull requests. It is [simple to set up](https://packit.dev/docs/packit-as-a-service/), free, and completely addresses point 3 above.
+
+Tests run on the [testing-farm](https://testing-farm.gitlab.io/api/), which provides reasonable (1 CPU, 2 GiB RAM) AWS EC2 instances for your project's tests. Critically, this is the *very same* infrastructure that is now being used for the Fedora gating tests.
```

Comment:
From the article, it's unclear if this the *very same* infrastructure by accident or on purpose. Hint it's the latter. Fedora gating tests run on the Testing Farm by design.

Reply:
Thanks! I added a sentence to  clarify this.